File: Python/BFSGraphMatrix.py
def bfs(graph, source, value):
    seen = []
    prev = []
    for _ in range(len(graph)):
        seen.append(False)
    for _ in range(len(graph)):
        prev.append(-1)
    
    seen[source] = True
    queue = [source]
    while len(queue) != 0:
        current = queue.pop(0)
        if current == value:
            break

        adjs = graph[current]
        for i in range(len(adjs)):
            if adjs[i] == 0:
                continue

            if seen[i]:
                continue

            seen[i] = True
            prev[i] = current
            queue.append(i)

    current = value
    out = []
    while prev[current] != -1:
        out.append(current)
        current = prev[current]
    
    if len(out) != 0:
        out.reverse()
        return [source] + out    
    return []

graph = [
    [0, 1, 1, 0, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 0, 1],
    [1, 0, 0, 0, 0],
    [0, 0, 0, 1, 0]
]
print(bfs(graph, 0, 4))

File: Java/BFSGraphMatrix.java
import java.util.*;
import java.util.Queue;

public class BFSGraphMatrix {
    public static LinkedList<Integer> bfs(int[][] graph, int source, int value){
        boolean[] seen = new boolean[graph.length];
        int[] prev = new int[graph.length];
        for(int i = 0; i < graph.length; i++){
            seen[i] = false;
        }
        for(int i = 0; i < graph.length; i++){
            prev[i] = -1;
        }

        seen[source] = true;
        Queue<Integer> queue = new LinkedList<>();
        queue.add(source);
        do{
           int current = queue.remove();
           if(current == value){
               break;
           }

           int[] adjs = graph[current];
           for(int i = 0; i < adjs.length; i++){
               if(adjs[i] == 0){
                   continue;
               }
               if(seen[i]){
                   continue;
               }
               seen[i] = true;
               prev[i] = current;
               queue.add(i);
           }
        } while(!queue.isEmpty());

        int current = value;
        LinkedList<Integer> out = new LinkedList<>();
        int i = 0;
        while(prev[current] != -1){
            out.add(current);
            current = prev[current];
            i++;
        }

        if(!out.isEmpty()){
            Collections.reverse(out);
            out.add(0, source);
            return out;
        }
        return new LinkedList<>();
    }

    public static void main(String[] args){
        int[][] graph = new int[][]{{0, 1, 1, 0, 0}, {0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}, {1, 0, 0, 0, 0}, {0, 0, 0, 1, 0}};
        System.out.println(bfs(graph, 0, 4));
    }
}
